index.js

```diff
@@ -39,116 +33,113 @@ function fromGeojsonVt (layers) {
- * serializer by encoding the feature geometry and properties.
- */
-function prepareLayer (layer) {
-  var preparedLayer = {
-    name: layer.name || '',
-    version: layer.version || 1,
-    extent: layer.extent || 4096,
-    keys: [],
-    values: [],
-    features: []
+function command (cmd, length) {
+  return (length << 3) + (cmd & 0x7)
+}
+
+function zigzag (num) {
+  return (num << 1) ^ (num >> 31)
+}
+
+function writeTile (tile, pbf) {
+  for (var key in tile.layers) {
+    pbf.writeMessage(3, writeLayer, tile.layers[key])
   }
+}
 
+function writeLayer (layer, pbf) {
+  pbf.writeVarintField(15, layer.version || 1)
+  pbf.writeStringField(1, layer.name || '')
+  pbf.writeVarintField(5, layer.extent || 4096)
+
+  var i
+  var keys = []
+  var values = []
   var keycache = {}
   var valuecache = {}
 
-  for (var i = 0; i < layer.length; i++) {
-    var feature = layer.feature(i)
-    feature.geometry = encodeGeometry(feature.loadGeometry())
-
-    var tags = []
-    for (var key in feature.properties) {
-      var keyIndex = keycache[key]
-      if (typeof keyIndex === 'undefined') {
-        preparedLayer.keys.push(key)
-        keyIndex = preparedLayer.keys.length - 1
-        keycache[key] = keyIndex
-      }
-      var value = wrapValue(feature.properties[key])
-      var valueIndex = valuecache[value.key]
-      if (typeof valueIndex === 'undefined') {
-        preparedLayer.values.push(value)
-        valueIndex = preparedLayer.values.length - 1
-        valuecache[value.key] = valueIndex
-      }
-      tags.push(keyIndex)
-      tags.push(valueIndex)
+  function writeFeature (feature, pbf) {
+    if (feature.id !== undefined) {
+      pbf.writeVarintField(1, feature.id)
     }
 
-    feature.tags = tags
-    preparedLayer.features.push(feature)
+    pbf.writeMessage(2, function () {
+      for (var key in feature.properties) {
+        var keyIndex = keycache[key]
+        if (typeof keyIndex === 'undefined') {
+          keys.push(key)
+          keyIndex = keys.length - 1
+          keycache[key] = keyIndex
+        }
+        pbf.writeVarint(keyIndex)
+
+        var value = feature.properties[key]
+        var type = typeof value
+        if (type !== 'string' && type !== 'boolean' && type !== 'number') {
+          value = JSON.stringify(value)
+        }
+        var valueKey = type + ':' + value
+        var valueIndex = valuecache[valueKey]
+        if (typeof valueIndex === 'undefined') {
+          values.push(value)
+          valueIndex = values.length - 1
+          valuecache[valueKey] = valueIndex
+        }
+        pbf.writeVarint(valueIndex)
+      }
+    })
+
+    pbf.writeVarintField(3, feature.type)
+
+    pbf.writeMessage(4, function () {
+      var x = 0
+      var y = 0
+      var geometry = feature.loadGeometry()
+      var rings = geometry.length
+      for (var r = 0; r < rings; r++) {
+        var ring = geometry[r]
+        pbf.writeVarint(command(1, 1)) // moveto
```

Comment:
See https://github.com/mapbox/vt-pbf/commit/21f0c8bf5c6affa375839a7d185d42f1983795c4#diff-168726dbe96b3ce427e7fedce31bb0bcR111, which I just merged before I noticed this PR.